pkg/controller/containerjfr/containerjfr_controller.go

```diff
@@ -95,182 +90,72 @@ func (r *ReconcileContainerJFR) Reconcile(request reconcile.Request) (reconcile.
-							Value: "8181",
-						},
-						{
-							Name:  "CONTAINER_JFR_EXT_WEB_PORT",
-							Value: "80",
-						},
-						{
-							Name: "CONTAINER_JFR_WEB_HOST",
-							ValueFrom: &corev1.EnvVarSource{
-								ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
-									LocalObjectReference: corev1.LocalObjectReference{
-										Name: "containerjfr",
-									},
-									Key: "CONTAINER_JFR_WEB_HOST",
-								},
-							},
-						},
-						{
-							Name:  "CONTAINER_JFR_LISTEN_PORT",
-							Value: "9090",
-						},
-						{
-							Name:  "CONTAINER_JFR_EXT_LISTEN_PORT",
-							Value: "80",
-						},
-						{
-							Name: "CONTAINER_JFR_LISTEN_HOST",
-							ValueFrom: &corev1.EnvVarSource{
-								ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
-									LocalObjectReference: corev1.LocalObjectReference{
-										Name: "containerjfr-command",
-									},
-									Key: "CONTAINER_JFR_LISTEN_HOST",
-								},
-							},
-						},
-					},
-				},
-			},
-		},
+func (r *ReconcileContainerJFR) createService(ctx context.Context, controller *rhjmcv1alpha1.ContainerJFR, svc *corev1.Service) error {
+	if err := controllerutil.SetControllerReference(controller, svc, r.scheme); err != nil {
+		return err
 	}
-}
-
-func newExporterServiceForPod(cr *rhjmcv1alpha1.ContainerJFR) *corev1.Service {
-	return &corev1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name,
-			Namespace: cr.Namespace,
-			Labels: map[string]string{
-				"app": cr.Name,
-			},
-			Annotations: map[string]string{
-				"fabric8.io/expose": "true",
-			},
-		},
-		Spec: corev1.ServiceSpec{
-			Type: corev1.ServiceTypeClusterIP,
-			Selector: map[string]string{
-				"app": cr.Name,
-			},
-			Ports: []corev1.ServicePort{
-				{
-					Name:       "8181-tcp",
-					Port:       8181,
-					TargetPort: intstr.IntOrString{IntVal: 8181},
-				},
-				{
-					Name:       "9091-tcp",
-					Port:       9091,
-					TargetPort: intstr.IntOrString{IntVal: 9091},
-				},
-			},
-		},
+	if err := r.createObjectIfNotExists(context.TODO(), types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace}, &corev1.Service{}, svc); err != nil {
+		return err
 	}
+	return nil
 }
 
-func newCommandChannelServiceForPod(cr *rhjmcv1alpha1.ContainerJFR) *corev1.Service {
-	return &corev1.Service{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      cr.Name + "-command",
-			Namespace: cr.Namespace,
-			Labels: map[string]string{
-				"app": cr.Name,
-			},
-			Annotations: map[string]string{
-				"fabric8.io/expose": "true",
-			},
-		},
-		Spec: corev1.ServiceSpec{
-			Type: corev1.ServiceTypeClusterIP,
-			Selector: map[string]string{
-				"app": cr.Name,
-			},
-			Ports: []corev1.ServicePort{
-				{
-					Name:       "9090-tcp",
-					Port:       9090,
-					TargetPort: intstr.IntOrString{IntVal: 9090},
-				},
-			},
-		},
+func (r *ReconcileContainerJFR) createObjectIfNotExists(ctx context.Context, ns types.NamespacedName, found runtime.Object, toCreate runtime.Object) error {
```

Comment:
It's used by `createService` with a `&corev1.Service{}` and in `Reconcile` when creating the pod (with a `corev1.Pod{}`), in either case so that the struct needed is passed to the `r.client.Get` function. I think this is required since without that type information, it can be ambiguous what kind of resource is being requested, since resources of different types within the same namespace can share the same name.

Reply:
Oh right, understood.